File: android/app/src/main/java/com/superproductivity/plugins/webdavhttp/WebDavHttpPlugin.java
package com.superproductivity.plugins.webdavhttp;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Iterator;

@CapacitorPlugin(name = "WebDavHttp")
public class WebDavHttpPlugin extends Plugin {

    private static final String[] WEBDAV_METHODS = {
        "PROPFIND", "PROPPATCH", "MKCOL", "COPY", "MOVE", "LOCK", "UNLOCK"
    };

    private boolean isWebDavMethod(String method) {
        for (String webdavMethod : WEBDAV_METHODS) {
            if (webdavMethod.equalsIgnoreCase(method)) {
                return true;
            }
        }
        return false;
    }

    @PluginMethod
    public void request(PluginCall call) {
        String urlString = call.getString("url");
        String method = call.getString("method", "GET");
        JSObject headers = call.getObject("headers", new JSObject());
        String data = call.getString("data");

        if (urlString == null) {
            call.reject("URL is required");
            return;
        }

        try {
            URL url = new URL(urlString);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();

            // For WebDAV methods, we need to use reflection since HttpURLConnection
            // only supports standard HTTP methods
            if (isWebDavMethod(method)) {
                try {
                    // First set it as POST to allow output
                    connection.setRequestMethod("POST");
                    // Then use reflection to set the actual method
                    java.lang.reflect.Field methodField = HttpURLConnection.class.getDeclaredField("method");
                    methodField.setAccessible(true);
                    methodField.set(connection, method);
                    
                    // Also update the delegate if using OkHttp
                    Object delegate = connection;
                    try {
                        java.lang.reflect.Field delegateField = connection.getClass().getDeclaredField("delegate");
                        delegateField.setAccessible(true);
                        delegate = delegateField.get(connection);
                        if (delegate != null) {
                            java.lang.reflect.Field delegateMethodField = delegate.getClass().getSuperclass().getDeclaredField("method");
                            delegateMethodField.setAccessible(true);
                            delegateMethodField.set(delegate, method);
                        }
                    } catch (Exception e) {
                        // OkHttp delegate not present, that's OK
                    }
                } catch (Exception e) {
                    call.reject("Failed to set WebDAV method: " + method, e);
                    return;
                }
            } else {
                // Standard HTTP method
                connection.setRequestMethod(method);
            }

            // Set headers
            Iterator<String> keys = headers.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                String value = headers.getString(key);
                connection.setRequestProperty(key, value);
            }

            // Handle request body
            if (data != null && !data.isEmpty()) {
                connection.setDoOutput(true);
                DataOutputStream outputStream = new DataOutputStream(connection.getOutputStream());
                outputStream.writeBytes(data);
                outputStream.flush();
                outputStream.close();
            }

            // Get response
            int responseCode = connection.getResponseCode();

            // Read response body
            BufferedReader reader;
            if (responseCode >= 200 && responseCode < 300) {
                reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            } else {
                reader = new BufferedReader(new InputStreamReader(connection.getErrorStream()));
            }

            StringBuilder responseBody = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                responseBody.append(line).append("\n");
            }
            reader.close();

            // Get response headers
            JSObject responseHeaders = new JSObject();
            for (String headerKey : connection.getHeaderFields().keySet()) {
                if (headerKey != null) {
                    responseHeaders.put(headerKey, connection.getHeaderField(headerKey));
                }
            }

            // Return response
            JSObject result = new JSObject();
            result.put("data", responseBody.toString());
            result.put("status", responseCode);
            result.put("headers", responseHeaders);
            result.put("url", connection.getURL().toString());

            call.resolve(result);

        } catch (Exception e) {
            call.reject("Request failed", e);
        }
    }
}

File: android/app/src/main/java/com/superproductivity/plugins/webdavhttp/WebDavHttpPlugin.kt
package com.superproductivity.plugins.webdavhttp

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLHandshakeException

@CapacitorPlugin(name = "WebDavHttp")
class WebDavHttpPlugin : Plugin() {

    companion object {
        // Shared OkHttpClient instance for better resource management
        private val client = OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            // Follow redirects as the TypeScript code expects
            .followRedirects(true)
            .followSslRedirects(true)
            // Add retry interceptor for reliability
            .addInterceptor { chain ->
                val request = chain.request()
                var response: Response? = null
                var tryCount = 0
                val maxRetries = 2
                
                while (tryCount < maxRetries) {
                    try {
                        response = chain.proceed(request)
                        break
                    } catch (e: IOException) {
                        tryCount++
                        if (tryCount >= maxRetries) {
                            throw e
                        }
                        // Wait before retry
                        Thread.sleep(1000L * tryCount)
                    }
                }
                
                response ?: throw IOException("Failed after $maxRetries retries")
            }
            .build()
    }

    @PluginMethod
    fun request(call: PluginCall) {
        // Execute in background to avoid blocking the UI thread
        GlobalScope.launch(Dispatchers.IO) {
            try {
                val result = performRequest(call)
                withContext(Dispatchers.Main) {
                    call.resolve(result)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    handleError(call, e)
                }
            }
        }
    }

    private fun performRequest(call: PluginCall): JSObject {
        val urlString = call.getString("url")
            ?: throw IllegalArgumentException("URL is required")
        val method = call.getString("method") ?: "GET"
        val headers = call.getObject("headers") ?: JSObject()
        val data = call.getString("data")

        // Build request
        val requestBuilder = Request.Builder()
            .url(urlString)

        // Add headers
        headers.keys().forEach { key ->
            headers.getString(key)?.let { value ->
                // Skip content-length header as OkHttp will set it
                if (!key.equals("content-length", ignoreCase = true)) {
                    requestBuilder.addHeader(key, value)
                }
            }
        }

        // Handle request body
        val requestBody: RequestBody? = when {
            data.isNullOrEmpty() && methodRequiresBody(method) -> {
                // Some WebDAV methods require a body even if empty
                "".toRequestBody()
            }
            !data.isNullOrEmpty() -> {
                val contentType = headers.getString("Content-Type") 
                    ?: headers.getString("content-type") 
                    ?: "application/xml; charset=utf-8"
                data.toRequestBody(contentType.toMediaType())
            }
            else -> null
        }

        // Set method with body
        when (method.uppercase()) {
            "GET" -> requestBuilder.get()
            "POST" -> requestBuilder.post(requestBody ?: "".toRequestBody())
            "PUT" -> requestBuilder.put(requestBody ?: "".toRequestBody())
            "DELETE" -> {
                if (requestBody != null) {
                    requestBuilder.delete(requestBody)
                } else {
                    requestBuilder.delete()
                }
            }
            "HEAD" -> requestBuilder.head()
            "PATCH" -> requestBuilder.patch(requestBody ?: "".toRequestBody())
            "OPTIONS" -> requestBuilder.method("OPTIONS", null)
            // WebDAV methods
            "PROPFIND", "PROPPATCH", "MKCOL", "COPY", "MOVE", "LOCK", "UNLOCK", "REPORT" -> {
                requestBuilder.method(method.uppercase(), requestBody)
            }
            else -> {
                throw IllegalArgumentException("Unsupported HTTP method: $method")
            }
        }

        val request = requestBuilder.build()

        // Execute request
        client.newCall(request).execute().use { response ->
            // Get response headers
            val responseHeaders = JSObject()
            response.headers.forEach { (name, value) ->
                // Convert header names to lowercase for consistency with web fetch
                responseHeaders.put(name.lowercase(), value)
            }

            // Get response body
            val responseBody = try {
                response.body?.string() ?: ""
            } catch (e: Exception) {
                // If body reading fails, return empty string
                ""
            }

            // Handle non-success status codes by including error info
            if (!response.isSuccessful) {
                // Log the error for debugging
                println("WebDavHttp: HTTP ${response.code} for $urlString")
            }

            // Return response matching the expected format
            return JSObject().apply {
                put("data", responseBody)
                put("status", response.code)
                put("headers", responseHeaders)
                put("url", response.request.url.toString())
            }
        }
    }

    private fun methodRequiresBody(method: String): Boolean {
        return when (method.uppercase()) {
            "POST", "PUT", "PATCH", "PROPFIND", "PROPPATCH", "REPORT", "LOCK" -> true
            else -> false
        }
    }

    private fun handleError(call: PluginCall, e: Exception) {
        when (e) {
            is UnknownHostException -> {
                call.reject("Network error: Unable to resolve host", "NETWORK_ERROR", e)
            }
            is SocketTimeoutException -> {
                call.reject("Network error: Request timeout", "TIMEOUT_ERROR", e)
            }
            is SSLHandshakeException -> {
                call.reject("SSL error: ${e.message}", "SSL_ERROR", e)
            }
            is IOException -> {
                call.reject("Network error: ${e.message}", "NETWORK_ERROR", e)
            }
            is IllegalArgumentException -> {
                call.reject(e.message ?: "Invalid argument", "INVALID_ARGUMENT", e)
            }
            else -> {
                call.reject("Unexpected error: ${e.message}", "UNKNOWN_ERROR", e)
            }
        }
    }
}